qurro/tests/test_qarcoal.py

```diff
@@ -0,0 +1,178 @@
+            qarcoal(get_mp_data.table, get_mp_data.taxonomy, num, denom)
+        assert "numerator not found!" == str(excinfo.value)
+
+    def test_invalid_denom(self, get_mp_data):
+        num = "Firm"
+        denom = "beyblade"
+        with pytest.raises(ValueError) as excinfo:
+            qarcoal(get_mp_data.table, get_mp_data.taxonomy, num, denom)
+        assert "denominator not found!" == str(excinfo.value)
+
+    def test_both_invalid(self, get_mp_data):
+        num = "beyblade"
+        denom = "yugioh"
+        with pytest.raises(ValueError) as excinfo:
+            qarcoal(get_mp_data.table, get_mp_data.taxonomy, num, denom)
+        assert "neither feature found!" == str(excinfo.value)
+
+    def test_shared_features_disallowed(self, get_mp_data):
+        num = "Firmicutes"
+        denom = "Bacilli"
+        with pytest.raises(ValueError) as excinfo:
+            qarcoal(
+                get_mp_data.table,
+                get_mp_data.taxonomy,
+                num,
+                denom,
+                allow_shared_features=False,
+            )
+        assert "Shared features" in str(excinfo.value)
+
+
+class TestOptionalParams:
+    def test_shared_features_allowed(self, get_mp_data):
+        num = "Firmicutes"
+        denom = "Bacilli"
+        qarcoal(
+            get_mp_data.table,
+            get_mp_data.taxonomy,
+            num,
+            denom,
+            allow_shared_features=True,
+        )
+
+    def test_samples_to_use(self, get_mp_data):
+        metadata_url = os.path.join(MP_URL, "sample-metadata.tsv")
+        sample_metadata = pd.read_csv(
+            metadata_url, sep="\t", index_col=0, skiprows=[1], header=0
+        )
+        gut_samples = sample_metadata[sample_metadata["BodySite"] == "gut"]
+        num_gut_samples = gut_samples.shape[0]
+        gut_samples = Metadata(gut_samples)
+
+        num = "p__Bacteroidetes"
+        denom = "p__Firmicutes"
+        q = qarcoal(
+            get_mp_data.table,
+            get_mp_data.taxonomy,
+            num,
+            denom,
+            samples_to_use=gut_samples,
+        )
+        assert q.shape[0] == num_gut_samples
+
+
+def test_large_numbers():
+    # Qurro fails when x > |2^53 - 1| due to JS implementation
+    bignum = 2 ** 53 - 1
+    large_vals = [np.random.randint(bignum + 1, bignum * 2) for x in range(30)]
+    mat = np.reshape(large_vals, (6, 5))
+    samps = ["S{}".format(i) for i in range(5)]
+    feats = ["F{}".format(i) for i in range(6)]
+    table = biom.table.Table(mat, feats, samps)
+    tax_labels = [
+        "Bulbasaur",
+        "Ivysaur",
+        "Venusaur",
+        "Charmander",
+        "Charmeleon",
+        "Charizard",
+    ]
+    confidence = ["0.99"] * 6
+    taxonomy = pd.DataFrame([feats, tax_labels, confidence]).T
+    taxonomy.columns = ["feature-id", "Taxon", "Confidence"]
+    taxonomy.set_index("feature-id", inplace=True, drop=True)
+    qarcoal(table, Metadata(taxonomy), "Char", "saur")
```

Comment:
Doesn't look like this checks the actual log-ratios -- would suggest adding that, if possible

Reply:
Sure - what kind of check do you think it should be? Just that they are valid (maybe this ties in to the previous issue of `SampleData[LogRatios]` validation (or lack there-of right now))

Reply:
Hm -- reading through this again, maybe it'd be best to just specify the `large_vals` manually and then verify that the log-ratios are (approximately) accurate based on a ground-truth. I like the idea of randomly generating large numbers to test, but being explicit would let us verify that log-ratios involving big numbers are properly handled.

Although JS does start to get wonky when you go above `Number.MAX_SAFE_INTEGER` (or below `Number.MIN_SAFE_INTEGER` I guess), it fails silently -- so if you were just running tests to make sure the code didn't crash (analogous the test here), you wouldn't actually notice an error unless you checked the output values you were getting. So, tl;dr... In order to prove that the Python code can handle essentially arbitrarily large numbers, I think we need to check the output values accordingly, preferably by specifying the desired output value manually.

That being said, we definitely don't need to check this with a 6x5 table (I don't want to ask you to manually come up with 30 different numbers!!!). I'd be perfectly content if this just checks, like, one or two log-ratios (maybe one involving numbers > 2**53 - 1, and one involving numbers < -(2**53 - 1)). The key here is that the test verifies that the Qarcoal output matches a known desired output value.

re: validation, not sure this sort of check would be doable there since I don't think the `SampleData[LogRatios]` type would have access on its own to the input feature table... that's going to be tricky to think about.

wow that turned into a massive amount of text, sorry! i can try to explain this more clearly tomorrow.